keps/sig-node/1664-node-ips/README.md

```diff
@@ -0,0 +1,722 @@
+then IPv6 DNS/interface addresses will be preferred; otherwise IPv4
+addresses will be preferred.
+
+#### New: `Node.Status.NodeIPs` and `--node-ips`
+
+To allow more consistent behavior without breaking any existing users,
+we will deprecate `--node-ip`, and deprecate the use of
+`Node.Status.Addresses` for determining the "primary" node IP. A new
+`--node-ips` argument will replace `--node-ip`, and a new
+`Node.Status.NodeIPs` field will provide an explicit indication of
+primary node IPs for components that were previously looking at
+`Node.Status.Addresses`.
+
+The new `--node-ips` argument can be either a single element
+describing the node IP configuration for a single family (indicating a
+mandatorily-single-stack node), or else a pair of elements describing
+both the IPv4 and IPv6 configurations. An element can be:
+
+  - a specific IP address, eg, `10.0.0.1` or `fd01::1`, indicating
+    that the node _must_ have that IP address for that IP family
+
+  - the string "`ipv4?`" or "`ipv6?`", indicating that the node _may_
+    have an IP address of that family, but is not required to
+
+  - the string "`ipv4`" or "`ipv6`", indicating that the node _must_
+    have an IP address of that family
+
+The default value is `ipv4?,ipv6?`, meaning that both IPv4 and IPv6
+addresses are desired, but not required, and that the IPv4 address
+should come first if both are available. (Other than the
+possible-dual-stack-ness, this corresponds to the current behavior of
+kubelet when `--node-ip` is not specified.) Depending on the available
+addresses, a node with this `--node-ips` value could end up with a
+single IPv4 IP, a single IPv6 IP, or a pair of IPs, IPv4 first, then
+IPv6.
+
+(Some other possibilities: `ipv4,ipv6` means mandatory dual-stack,
+preferring IPv4. Either `ipv4` or `ipv4?` indicate mandatory
+single-stack IPv4. (Technically `ipv4?` means "an IPv4 address is
+desired (and no other address type is allowed)", which has the same
+effect as "an IPv4 address is required".) Meanwhile, `ipv4,ipv6?`
+allows either single-stack IPv4 or dual-stack preferring IPv4, while
+`ipv6?,ipv4` allows either single-stack IPv4 or dual-stack preferring
+IPv6.)
+
+If `--node-ips` is not specified but `--node-ip` is, then `--node-ips`
+is defaulted to an equivalent value as follows:
+
+  - If `--node-ip` is `0.0.0.0` then `--node-ips` is set to
+    `ipv4?,ipv6?`.
+
+  - If `--node-ip` is `::` then `--node-ips` is set to `ipv6?,ipv4?`.
+
+  - Otherwise (when `--node-ip` is a specific IP address) `--node-ips`
+    is set to the same value as `--node-ip`.
+
+That is, unless you specify a specific IP, it will potentially fall
+back to the "wrong" IP family, just like with current kubelet.
+
+It is an error to pass both `--node-ip` and `--node-ips`.
+
+Via complicated means (discussed under
+[Implementation](#implementation)), `Node.Status.NodeIPs` will then
+end up with 1 or 2 node IPs. Even in clusters with `IPv6DualStack`
```

Comment:
I foresee that in bare metal some complex scenarios will ask for having much more IPs, is this 2 IPs a hard limit or just a first step to move on with dual stack?

Reply:
I said not to read that version! :slightly_smiling_face: 

Earlier discussion was leaning more toward "we have too many node addresses that we don't need" more than "we don't have enough node addresses". Note that this doesn't mean that nodes can't actually have more IPs; it's just that those IPs won't be reflected in the Node object.

At any rate, if people decide there is some reason to be able to create additional NodeAddresses on bare metal in the future, then that can be added in the future. It is not needed for dual-stack now.